File: tests/exchange_nft_for_tokens.ts
import * as anchor from "@coral-xyz/anchor";
import { Program } from "@coral-xyz/anchor";
import { MintSplToken } from "../target/types/mint_spl_token";
import {
    PublicKey,
    SystemProgram,
    Keypair
} from "@solana/web3.js";
import {
    TOKEN_PROGRAM_ID,
    ASSOCIATED_TOKEN_PROGRAM_ID,
    getAssociatedTokenAddressSync,
    createMint, 
    mintTo,     
    getAccount, 
    getMint
} from "@solana/spl-token";
import { assert } from "chai"; 

describe("nft_exchange (Local Test with Provider Wallet)", () => {
    const provider = anchor.AnchorProvider.env();
    anchor.setProvider(provider);
    const program = anchor.workspace.MintSplToken as Program<MintSplToken>;
    const connection = provider.connection;

    const testWallet = provider.wallet as anchor.Wallet;
    console.log(`Using Test Wallet: ${testWallet.publicKey.toBase58()} as Investor & Mint Authority`);

    const amount = new anchor.BN(1); 
    const nftMintAddress = new PublicKey("2o5WswYePk6pq2YZ5dKxN79UyRZRw25sLM55oVSVnVho");
    const fungibleMintAddress = new PublicKey("DNMVCRJKfHe4yz5tcgteSn6e75WHXUtoZZxKjcii2uze");

    let walletNftAta: PublicKey;
    let walletFungibleAta: PublicKey;

    before(async () => {
        console.log("--- Test Setup ---");
        walletNftAta = getAssociatedTokenAddressSync(nftMintAddress, testWallet.publicKey);
        walletFungibleAta = getAssociatedTokenAddressSync(fungibleMintAddress, testWallet.publicKey);
        console.log(`Wallet NFT ATA: ${walletNftAta.toBase58()}`);
        console.log(`Wallet Fungible ATA: ${walletFungibleAta.toBase58()}`);

        // NFT Check
        try {
            const nftAccount = await getAccount(connection, walletNftAta);
            assert.strictEqual(nftAccount.amount.toString(), "1", `Wallet ${testWallet.publicKey.toBase58()} does not hold exactly 1 NFT from mint ${nftMintAddress.toBase58()}`);
            console.log(`Wallet already owns the required NFT in ATA ${walletNftAta.toBase58()}.`);
        } catch (e) {
            console.warn(`NFT ATA ${walletNftAta.toBase58()} not found or NFT is missing for wallet ${testWallet.publicKey.toBase58()}.`);
            console.warn(`You might need to mint the NFT (${nftMintAddress.toBase58()}) to this wallet manually or add minting logic to the 'before' block if the wallet is the NFT mint authority.`);
            throw new Error("NFT ownership check failed. Cannot proceed.");
            /*
            console.log("Attempting to mint NFT...");
            await mintTo(connection, testWallet.payer, nftMintAddress, walletNftAta, testWallet.publicKey, 1, [testWallet.payer], { commitment: "confirmed" });
            console.log("NFT minted for test setup.");
            */
        }
        console.log("--- Setup Complete ---");
    });

    it("Should exchange NFT for tokens (using provider wallet)", async () => {
        
        const nftBalanceBefore = (await getAccount(connection, walletNftAta)).amount;
        // let fungibleBalanceBefore = new anchor.BN(2);;
        // try { fungibleBalanceBefore = (await getAccount(connection, walletFungibleAta)).amount; } catch (e) {}
        // console.log(`Balances before: NFT=${nftBalanceBefore}, Fungible=${fungibleBalanceBefore}`);

        try {
            const tx = await program.methods
                .exchangeNftForTokens(amount)
                .accounts({
                    investor: testWallet.publicKey,
                    mintAuthority: testWallet.publicKey,

                    nftMint: nftMintAddress,
                    fungibleMint: fungibleMintAddress,
                })
                .signers([testWallet.payer]) 
                .rpc({ commitment: "confirmed" }); 
            try {
                const nftAccountInfo = await getAccount(connection, walletNftAta);
                assert.strictEqual(nftAccountInfo.amount.toString(), "0", "NFT account balance is not zero");
                console.log("NFT account balance verified: 0");
            } catch (error) {
                 if (error.message.includes("Account does not exist") || error.message.includes("could not find account")) {
                     console.log("NFT account closed as expected.");
                 } else { throw error; }
            }

             const fungibleAccountInfo = await getAccount(connection, walletFungibleAta);
             const fungibleMintInfo = await getMint(connection, fungibleMintAddress);
             const expectedAmount = amount.mul(new anchor.BN(10).pow(new anchor.BN(fungibleMintInfo.decimals)));
             console.log(`Fungible token balance verified: ${fungibleAccountInfo.amount}`);


        } catch (error) {
            console.error("Error during exchange:", error);
             if (error instanceof anchor.AnchorError) {
                 console.error("\nProgram Logs:", error.logs);
             }
            throw error; 
        }
    });
});
// For integratio with frontend
/*


describe("nft_exchange", () => {
    const provider = anchor.AnchorProvider.env();
    anchor.setProvider(provider);
    const program = anchor.workspace.mintSplToken as Program<MintSplToken>;
 const amount = new anchor.BN(1);
    const investorAddress = new PublicKey("7bCkmDRYMPTSYJPEhdx9rvWhQb3UpbBPDVuSDhWThPTF");
    const nftMintAddress = new PublicKey("4GTyiNebtcvtqVdUgS9SwZwK3x3m7rCvFag4AeM7NVzX");

    
    const fungibleMintAddress = new PublicKey("DNMVCRJKfHe4yz5tcgteSn6e75WHXUtoZZxKjcii2uze"); // ПРИКЛАД - ЗАМІНІТЬ!
    const mintAuthorityAddress = new PublicKey("HECVDSyH6doUTJ7FQ7K8pRRiQm3ojPAuPz63zH3Dep9G"); // ПРИКЛАД - ЗАМІНІТЬ!
    const fungibleAmount = new anchor.BN(1);
    const payer = provider.wallet;
    const investor = new PublicKey("7bCkmDRYMPTSYJPEhdx9rvWhQb3UpbBPDVuSDhWThPTF");
    let investorNftAta: PublicKey;
    let investorFungibleAta: PublicKey;

  it("Should mint tokens to associated account!", async () => { 
   
    let tx: Transaction | null = null; 

    try{
       tx = await program.methods
        .exchangeNftForTokens(
          amount,
        )
        .accounts({
            investor: investorAddress,
            nftMint: nftMintAddress,
            mintAuthority: mintAuthorityAddress,
            fungibleMint: fungibleMintAddress,
        })
        // .signers([mintKeypair])
        // .signers()
        .transaction()
        tx.feePayer = investorAddress;

        // .rpc();
        console.log("Your transaction signature", tx);
      }catch (error) {
        console.error("Error creating token: ", error);
      }
      console.log("SPL token was minted successfully!")
  });


});*/

File: programs/mint_spl_token/src/instructions/exchange_nft_for_tokens.rs
use {
    anchor_lang::prelude::*,
    anchor_spl::{
        associated_token::AssociatedToken,
        token::{mint_to, MintTo, TokenAccount, Mint, Token, burn, Burn}
    },
};

pub fn exchange_nft_for_tokens(
    ctx:Context<ExchangeNFTForTokens>,
    amount: u64,
) -> Result<()>{
    msg!("Minting tokens to associated token account...");
    msg!("Mint: {}", &ctx.accounts.fungible_mint.key());
    msg!(
        "Token Address: {}",
        &ctx.accounts.associated_token_account.key()
    );
    
    // 1. Burn NFT 
    msg!("Burning NFT...");

    burn(
        CpiContext::new(
            ctx.accounts.token_program.to_account_info(),
            Burn{
                mint: ctx.accounts.nft_mint.to_account_info(),
                from: ctx.accounts.nft_token_account.to_account_info(),
                authority: ctx.accounts.investor.to_account_info(),
            }, 
        ),  
        1, // NFT amount 
    )?;
    msg!("NFT burned successfully.");

    // 2. Mint fungible tokens 
    msg!("Minting fungible tokens...");
    mint_to(
        CpiContext::new(
            ctx.accounts.token_program.to_account_info(),
            MintTo{
                mint: ctx.accounts.fungible_mint.to_account_info(),
                to: ctx.accounts.associated_token_account.to_account_info(),
                authority: ctx.accounts.mint_authority.to_account_info(),
            },
        ),
        amount * 10u64.pow(ctx.accounts.fungible_mint.decimals as u32), // Mint tokens, adjust for decimals
    )?;
    msg!("Token minted successfully.");

    Ok(())
}


#[derive(Accounts)]
pub struct ExchangeNFTForTokens<'info>{
    #[account(mut)]
    pub mint_authority: Signer<'info>,
    #[account(mut)]
    pub fungible_mint:Account<'info, Mint>,
    #[account(mut)]
    pub investor: Signer<'info>,
    // це токен-акаунт отримувача для конкретного типу токена (mint_account)
    // mut
    #[account(
        init_if_needed,
        payer = investor,
        associated_token::mint = fungible_mint,
        associated_token::authority = investor,
    )]    
    pub associated_token_account: Account<'info, TokenAccount>,

   
    #[account(mut)]
    pub nft_mint: Account<'info, Mint>,
    // #[account(mut, has_one = investor)]
    #[account( 
        mut, // mut, бо зменшується amount і акаунт буде закрито
        associated_token::mint = nft_mint, // Перевірка, що ATA для правильного мінта NFT
        associated_token::authority = investor, // Перевірка, що власник ATA - це інвестор, що підписав
        constraint = nft_token_account.amount == 1 //@ ErrorCode::NftAccountEmptyOrWrongAmount
    )]
    pub nft_token_account: Account<'info, TokenAccount>,
    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub system_program:Program<'info, System>,
}
#[error_code]
pub enum ErrorCode {
    #[msg("NFT token account is empty or does not hold exactly one NFT.")]
    NftAccountEmptyOrWrongAmount,
    #[msg("Amount calculation overflowed.")]
    AmountOverflow,
}